examples/petsc/setup.h

```diff
@@ -0,0 +1,881 @@
+}
+
+// Set up libCEED for a given degree
+static int SetupLibceedByDegree(DM dm, Ceed ceed, CeedInt degree, CeedInt dim,
+    CeedInt qextra, PetscInt ncompu, PetscInt gsize, PetscInt xlsize,
+    bpType bpChoice, CeedData data, PetscBool setup_rhs,
+    CeedVector rhsceed, CeedVector *target) {
+  int ierr;
+  DM dmcoord;
+  PetscSection section;
+  Vec coords;
+  const PetscScalar *coordArray;
+  CeedBasis basisx, basisu;
+  CeedElemRestriction Erestrictx, Erestrictu, Erestrictxi,
+                      Erestrictui, Erestrictqdi;
+  CeedQFunction qf_setupgeo, qf_apply;
+  CeedOperator op_setupgeo, op_apply;
+  CeedVector xcoord, qdata, xceed, yceed;
+  CeedInt P, Q, cStart, cEnd, nelem, qdatasize = bpOptions[bpChoice].qdatasize;
+
+  // CEED bases
+  P = degree + 1;
+  Q = P + qextra;
+  CeedBasisCreateTensorH1Lagrange(ceed, dim, ncompu, P, Q,
+                                  bpOptions[bpChoice].qmode, &basisu);
+  CeedBasisCreateTensorH1Lagrange(ceed, dim, dim, 2, Q,
+                                  bpOptions[bpChoice].qmode, &basisx);
+
+  // CEED restrictions
+  ierr = DMGetCoordinateDM(dm, &dmcoord); CHKERRQ(ierr);
+  ierr = DMPlexSetClosurePermutationTensor(dmcoord, PETSC_DETERMINE, NULL);
+  CHKERRQ(ierr);
+
+  CreateRestrictionPlex(ceed, 2, dim, &Erestrictx, dmcoord);
+  CreateRestrictionPlex(ceed, P, ncompu, &Erestrictu, dm);
+
+  ierr = DMPlexGetHeightStratum(dm, 0, &cStart, &cEnd); CHKERRQ(ierr);
+  nelem = cEnd - cStart;
+
+  CeedElemRestrictionCreateIdentity(ceed, nelem, Q*Q*Q, nelem*Q*Q*Q, ncompu,
+                                    &Erestrictui); CHKERRQ(ierr);
+  CeedElemRestrictionCreateIdentity(ceed, nelem, Q*Q*Q, nelem*Q*Q*Q,
+                                    qdatasize, &Erestrictqdi); CHKERRQ(ierr);
+  CeedElemRestrictionCreateIdentity(ceed, nelem, Q*Q*Q, nelem*Q*Q*Q, dim,
+                                    &Erestrictxi); CHKERRQ(ierr);
+
+  // Element coordinates
+  ierr = DMGetCoordinatesLocal(dm, &coords); CHKERRQ(ierr);
+  ierr = VecGetArrayRead(coords, &coordArray); CHKERRQ(ierr);
+  ierr = DMGetSection(dmcoord, &section); CHKERRQ(ierr);
+
+  CeedElemRestrictionCreateVector(Erestrictx, &xcoord, NULL);
+  CeedVectorSetArray(xcoord, CEED_MEM_HOST, CEED_COPY_VALUES,
+                     (PetscScalar *)coordArray);
+  ierr = VecRestoreArrayRead(coords, &coordArray); CHKERRQ(ierr);
+
+  // Create the persistent vectors that will be needed in setup and apply
+  CeedInt nqpts;
+  CeedBasisGetNumQuadraturePoints(basisu, &nqpts);
+  CeedVectorCreate(ceed, qdatasize*nelem*nqpts, &qdata);
+  CeedVectorCreate(ceed, xlsize, &xceed);
+  CeedVectorCreate(ceed, xlsize, &yceed);
+
+  // Create the Q-function that builds the operator (i.e. computes its
+  // quadrature data) and set its context data
+  CeedQFunctionCreateInterior(ceed, 1, bpOptions[bpChoice].setupgeo,
+                              bpOptions[bpChoice].setupgeofname, &qf_setupgeo);
+  CeedQFunctionAddInput(qf_setupgeo, "dx", dim*dim, CEED_EVAL_GRAD);
```

Comment:
Sorry @jeremylt , I know this has been merged already and I'm getting late to this, but I didn't get to review this refactoring before being merged. We made an effort in PR #335 to distinguish all `dim*dim` in `ncompx*dim`, since these examples are intended to be like tutorials

Reply:
Yeah, I can change that pretty easily.